File: common/src/test/java/com/hartwig/actin/molecular/filter/GeneAggregatorTest.java
package com.hartwig.actin.molecular.filter;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.Set;

import com.hartwig.serve.datamodel.common.GeneRole;
import com.hartwig.serve.datamodel.gene.ImmutableKnownGene;

import org.jetbrains.annotations.NotNull;
import org.junit.Test;

public class GeneAggregatorTest {

    private static final String GENE = "gene";
    private static final ImmutableKnownGene ONCO_GENE = gene(GENE, GeneRole.ONCO);
    private static final ImmutableKnownGene UNKNOWN_GENE = gene(GENE, GeneRole.UNKNOWN);

    @Test
    public void shouldUseKnownGeneRolesWhenMultipleRoles() {
        assertThat(GeneAggregator.aggregate(Set.of(ONCO_GENE, UNKNOWN_GENE))).containsOnly(ONCO_GENE);
    }

    @Test
    public void shouldUseUnknownGeneRolesWhenOnlyRole() {
        assertThat(GeneAggregator.aggregate(Set.of(UNKNOWN_GENE))).containsOnly(UNKNOWN_GENE);
    }

    @Test
    public void shouldNotAggregateWhenNotRequired() {
        ImmutableKnownGene anotherGene = ONCO_GENE.withGene("another_gene");
        assertThat(GeneAggregator.aggregate(Set.of(ONCO_GENE, anotherGene))).containsOnly(ONCO_GENE, anotherGene);
    }

    @NotNull
    private static ImmutableKnownGene gene(String name, GeneRole role) {
        return ImmutableKnownGene.builder().gene(name).geneRole(role).build();
    }

}

File: molecular/src/test/kotlin/com/hartwig/actin/molecular/evidence/known/GeneAggregatorTest.kt
package com.hartwig.actin.molecular.evidence.known

import com.hartwig.serve.datamodel.common.GeneRole
import com.hartwig.serve.datamodel.gene.ImmutableKnownGene
import com.hartwig.serve.datamodel.gene.KnownGene
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

private const val GENE: String = "gene"

class GeneAggregatorTest {

    private val oncoGene = gene(GeneRole.ONCO)
    private val bothGene = gene(GeneRole.BOTH)
    private val unknownGene = gene(GeneRole.UNKNOWN)

    @Test
    fun `Should use both gene roles when onco, both and unknown in set`() {
        assertThat(GeneAggregator.aggregate(setOf(oncoGene, bothGene, unknownGene))).containsOnly(bothGene)
    }

    @Test
    fun `Should use onco gene role when onco and unknown in set`() {
        assertThat(GeneAggregator.aggregate(setOf(oncoGene, unknownGene))).containsOnly(oncoGene)
    }

    @Test
    fun `Should use unknown gene role when only role present`() {
        assertThat(GeneAggregator.aggregate(setOf(unknownGene))).containsOnly(unknownGene)
    }

    @Test
    fun `Should only aggregate when required`() {
        val anotherGene = ImmutableKnownGene.copyOf(oncoGene).withGene("another_gene")
        assertThat(GeneAggregator.aggregate(setOf(oncoGene, anotherGene))).containsOnly(oncoGene, anotherGene)
    }

    private fun gene(role: GeneRole): KnownGene {
        return ImmutableKnownGene.builder().gene(GENE).geneRole(role).build()
    }
}